Roman_Shagun_rshag17@gmail.com

mdalag opened the pull request.

mdalag committed:
First iteration without tests

Add exceptions.py for custom exceptions
Add logger.py for convenient output managment
Add rss_feed.py to represent whole rss rss feed
Add rss_item.py to store a single news item
Implement rss_reader.py
Add non-standart libs in requirements.txt

mdalag committed:
Iteration 1 complete

 - Add tests for RssFeed and RssItem classes.
 - Refactor rss_reader.init_feed(): add new func
to init list of news_items separately
 - Change RssFeed.print_feed() to be
more convenient for usage and testing
 - Add README

HenadziStantchik commented on final_task/rss_reader/classes/logger.py:
```diff
@@ -0,0 +1,19 @@
+        Receives message and prints it into stdout if logging is turned on
+        '''
+        if logger.is_log_on:
+            logging.debug(msg)
```
You do not really need class for this. More so if you have just one static method in it. Simple function will suffice.

mdalag replied:
Done

HenadziStantchik commented on final_task/rss_reader/classes/rss_item.py:
```diff
@@ -0,0 +1,39 @@
+        data = data.replace('&#34;', '"')
+        data = data.replace('&#x22;', '"')
+        data = data.replace('&#39;', '\'')
+        data = data.replace('&#x27;', '\'')
```
It is better to find a more universal solution is some libraries. There may be other symbols which have encoding problems.

mdalag replied:
Done

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,71 @@
+        tit = entry.get('title', 'unknown')
+        pub = entry.get('published', 'unknown')
+        lnk = entry.get('link', 'no link')
+        med = entry.get('media_content', 'no media content')
```
Please use more infromative variable names.

mdalag replied:
Done

HenadziStantchik commented on final_task/rss_reader/classes/rss_feed.py:
```diff
@@ -0,0 +1,46 @@
+            result += '='*120 + '\n'
+        print(result)
+    
+    def toJSON(self):
```
You may want to check your code with `pycodestyle` tool.
(optional - set a max line length to 120 while doing so)

mdalag replied:
Done

HenadziStantchik replied:
Also, function names should be in lowercase, preferably using `snake_case_notation`

HenadziStantchik commented on final_task/rss_reader/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+feedparser==5.2.1
+argparse==1.4.0
+json==2.0.9
```
json is a standart library, no need to include it in this file

mdalag replied:
Done

HenadziStantchik commented:
If the `--limit` is not specified, user should get all available feed. The same if the limit is too large.

mdalag committed:
Iteration 2 complete

 - Implement setup.py
 - Add installation guide to README
 - Refactor code to log without custom logger
 - Fix PEP8 issues in rss_feed.py
 - Refactor rss_item.py to decode symbols via html lib
 - Remove standart libs fom requirements.txt
 - Fix var name issues in rss_feed.py

HenadziStantchik commented on final_task/rss_reader/classes/exceptions.py:
```diff
@@ -0,0 +1,6 @@
+class FeedError(Exception):
+    '''
+    Raise when something is wrong with feed
+    Really I don't feel good about this class
```
This class is correct. Although it is better to have multiple custom exception classes for each specific error case

HenadziStantchik commented on final_task/rss_reader/classes/rss_feed.py:
```diff
@@ -0,0 +1,47 @@
+        self.news_list = news_list
+        self.title = title
+        self.description = description
+        self.link = link
```
Consider using a `dataclass` for this

UPD: This is optional, but looks like it is a logical thing to do. This class has only `__init__` without any complex logic and 2 methods that just print data in different representations

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,83 @@
+    level=logging.DEBUG,
+    format='[%(asctime)s] {%(filename)s} %(levelname)s - %(message)s'
+    )
+logger = logging.getLogger('rss_logger')
```
global variables usually declared in uppercase, e.g. `LOGGER`

mdalag replied:
Done

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,83 @@
+            if media_link == '':
+                media_link = 'no media'
+        news = RssItem(title, published, link, media_link)
+        news_list.append(news)
```
Consider using `slices` here instead of comparing index to limit for every iteration

mdalag replied:
Done

HenadziStantchik commented on final_task/setup.py:
```diff
@@ -0,0 +1,26 @@
+    from pip.req import parse_requirements
+
+with open('final_task/rss_reader/requirements.txt') as fp:
+    install_requires = fp.read()
```
It is better to manually add them, instead of reading a file here

mdalag replied:
Done

HenadziStantchik commented:
Your app does not work if user tries to run it without using setuptools. (`ModuleNotFoundError`)

mdalag committed:
Iteration 3 complete

 - Fix function names
 - Fix global variables names
 - Use slices in rss_reader.init_news_list
 - Implement manual adding of packages in setup.py
 - Change project structure to fix import errors
 - Change README
 - Add test for rss_reader.py
 - Add test test_feed.xml as test feed

HenadziStantchik commented on final_task/rss_reader/rss_item.py:
```diff
@@ -0,0 +1,36 @@
+    source: str
+    date: str
+
+    @classmethod
```
Why did you need to make this a `classmethod`?

mdalag replied:
This is a factory method which is called to create an instance of the class from dict

dzhigailo replied:
Why can't you just use RssItem(**item)? You dont really need this method

HenadziStantchik commented on final_task/rss_reader/rss_item.py:
```diff
@@ -0,0 +1,36 @@
+            media=item_dict['media'],
+            source=item_dict['source'],
+            date=item_dict['date']
+            )
```
Check out function `asdict` from `dataclass` lib. It is better to try to use it here instead

mdalag replied:
If I got it correct, `asdict` converts a class to dict, while `from_dict` creates the class from dict. I'm not sure that I could use `asdict` here

HenadziStantchik replied:
Yes, my bad. I misunderstood what method does. Still you can use unpacking here, instead of passing each argument by name

mdalag replied:
Will unpacking be suitable if I'm going to process some fields in this method?

HenadziStantchik replied:
Well, it depends. You can do something like this:
```
item_dict['field'] = do_some_stuff(item_dict['field'])
.....
return cls(**item_dict)
```
But if you are going to change many fields in the dict, then it is better to leave it as is. See for yourself and use what do you think is more readable

mdalag replied:
Ok, thanks

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,127 @@
+            get_news_by_date(None, args.date, args.limit)
+        else:
+            print('Positional argument \'url\' is required')
+            sys.exit(1)
```
It is better to have only one exit point from the application. For example you can raise your custom exception here and handle it lower

mdalag replied:
Got it

HenadziStantchik commented:
Here is some feedback on your app work:

1. Your app does not print `DESCRIPTION` field of news entry. (check the example in FinalTAsk.md)

Keep in mind that it should not have encoding problems too.

2. `--date` does not work together with `--json`

mdalag committed:
Iteration 4

 - Fix multiple exit points
 - Add DESCRIPTION
 - Fix --date now working with --json
 - Refactor rss_reader to be more clear
 - Add converters to .pdf and .html

HenadziStantchik commented on final_task/rss_reader/converters.py:
```diff
@@ -0,0 +1,170 @@
+
+import requests
+from dominate import document
+from dominate.tags import *
```
It is better to avoid wildcard import whenever you can

mdalag replied:
Done

HenadziStantchik commented on final_task/rss_reader/converters.py:
```diff
@@ -0,0 +1,170 @@
+        path += 'feed-' + str(datetime.datetime.now()) + get_extention(mode)
+    else:
+        path += '/feed-' + str(datetime.datetime.now()) + get_extention(mode)
+    try:
```
For all operations with filesystem paths it is better to use `os.path` module for crossplatform applications.

mdalag replied:
Fixed

HenadziStantchik commented on final_task/rss_reader/rss_feed.py:
```diff
@@ -0,0 +1,77 @@
+import json
+import logging
+import re
+import sys
```
Unused imports here

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,206 @@
+import logging
+import re
+import sys
+
```
unused import here

mdalag replied:
Fixed

dzhigailo commented on .coverage:
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/home/mdalag/Documents/epam_task/FinalTaskRssParser/final_task/rss_reader/converters.py":[1,2,3,5,6,7,8,10,12,13,15,19,20,21,22,23,24,25,26,27,28,29,30,31,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,76,78,87,88,89,90,91,92,93,96,158],"/home/md ...
```
Please, add this file and file below to .gitignore

dzhigailo commented on final_task/rss_reader/.coverage:
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/usr/local/lib/python3.7/dist-packages/rss_reader-1.0-py3.7.egg/tests/test_rss_item.py":[1,2,3,4,6,7,10,12],"/home/mdalag/Documents/epam_task/FinalTaskRssParser/final_task/rss_reader/rss_item.py":[1,2,3,4,5,8,17,18,19,20,21,22,23,24,25,27,32,33,45,52,62],"/home/mdalag/.local/lib/python3.7/site-packages/jsonpickle/__init__.p ...
```
All .coverage related files should be added to .gitignore

mdalag replied:
Done

dzhigailo commented on final_task/rss_reader/converters.py:
```diff
@@ -0,0 +1,170 @@
+        path += '/feed-' + str(datetime.datetime.now()) + get_extention(mode)
+    try:
+        LOGGER.debug('TRYING TO CREATE FILE...')
+        export_file = open(path, 'w', encoding='utf-8')
```
'w' flag in open function will always create new file if one not exists or overwrite existing file. If you want to check if path is valid, you could use tools from os lib
Also, please always use context manager while working with files.

mdalag replied:
Got it

HenadziStantchik commented:
1. Please ensure that you have no encoding problems in converted html file.

Aside from this, i did not discover any issues with Iteration 4 work

dzhigailo commented on final_task/rss_reader/converters.py:
```diff
@@ -0,0 +1,170 @@
+        pdf.set_x(4)
+        pdf.cell(20, 6, 'Description:', ln=1)
+        pdf.set_x(20)
+        if len(news_item['description']) == 0:
```
`
if not news_item['description']:
`

mdalag replied:
Fixed

mdalag committed:
Fix iteration 4

 - Fix indicated errors
 - Use pkg_resources for dynamic font detection
 - Change .pdf font to handle more symbols
 - Add font to MANIFEST.in
 - Fix unescaped characters in json and html
 - Change README.md setup section
 - Add tests for rss_reader

HenadziStantchik commented:
I did not discover any issues with Iteration 4 functionality, but be sure to test it thoroughly in case I missed something

mdalag committed:
Iteration 5

- Add colored output for news feed via colorama

mdalag committed:
Fix path to file in converters.py according to OS

 - Convertion now works on Windows

HenadziStantchik commented on final_task/README.md:
```diff
@@ -1,3 +1,90 @@
+# Quick start
+
+## Usage
+    usage: rss_reader.py [-h] [--version] [--json] [--verbose] [--limit LIMIT] [--date DATE] [source]
+
+    RRS feed receiver
+
+    positional arguments:
+    source            URL for RSS feed
+
+    optional arguments:
+    -h, --help         show this help message and exit
+    --version          prints version
+    --json             converts news to JSON
+    --verbose          output verbose status messages
+    --limit LIMIT      determines the number of showed news.
+    --date DATE        shows cached news at given date
+    --to_pdf TO_PDF    coverts news to PDF.
+    --to_html TO_HTML  coverts news to HTML.
+
+    TO_PDF/TO_HTML - path to directory for file
+    File's name is in format feed-*current datetime*.*extention*
```
It seems you forgot to add `--colorize` here.

HenadziStantchik commented on final_task/rss_reader/converters.py:
```diff
@@ -0,0 +1,191 @@
+    return temp_img
+
+
+def get_pdf_doc(news_list):
```
This function is too large, it is better to try to split it.

HenadziStantchik commented on final_task/tests/readme.txt:
```diff
@@ -0,0 +1,3 @@
+Some shit happend :c with imports and I was had to run tests from /FinalTaskRssParser dir like this
```
Please do not use such language on work in any form.

HenadziStantchik commented:
Here is feedback on your application:
Iteration 1:
Everything works fine.
Iteration 2:
Installs fine, all iterations work.
Iteration 3:
`--date` should be specified in `YYYYMMDD` format, not `YYYYMMD`:
![image](https://user-images.githubusercontent.com/43743521/70061910-27144780-15f6-11ea-8038-58b465cd9098.png)
Iteration 4:
Everything works fine.
Iteration 5:
`--colorize` must be just an option.



Tests are low in number, but the coverage percentage is enough. Still this is not a really good thing.
Code is mostly readable, absolutely corresponds to the PEP8 guidelines.
Commit messages are very informative and absolutely correspond to the recommended commit message style.
